crates/astria-sequencer/src/transaction/signed.rs

```diff
@@ -0,0 +1,127 @@
+use anyhow::{
+    bail,
+    Context as _,
+    Result,
+};
+use astria_proto::sequencer::v1::{
+    unsigned_transaction::Value::AccountsTransaction as ProtoAccountsTransaction,
+    SignedTransaction as ProtoSignedTransaction,
+};
+use async_trait::async_trait;
+use penumbra_storage::{
+    StateRead,
+    StateWrite,
+};
+use prost::Message as _;
+use tracing::instrument;
+
+use crate::{
+    accounts::{
+        transaction::Transaction as AccountsTransaction,
+        types::Address,
+    },
+    crypto::{
+        Signature,
+        VerificationKey,
+    },
+    transaction::{
+        unsigned::Transaction as UnsignedTransaction,
+        ActionHandler,
+    },
+};
+
+/// Represents a transaction signed by a user.
+/// It contains the signature and the public key of the user,
+/// as well as the transaction itself.
+#[derive(Debug, Clone)]
+pub(crate) struct Transaction {
+    pub(crate) signature: Signature,
+    pub(crate) public_key: VerificationKey,
+    pub(crate) transaction: UnsignedTransaction,
+}
+
+impl Transaction {
+    /// Verifies the transaction signature.
+    /// The transaction signature message is the hash of the transaction.
+    ///
+    /// # Errors
+    ///
+    /// - If the signature is invalid
+    pub(crate) fn verify_signature(&self) -> Result<()> {
+        self.public_key
+            .verify(&self.signature, &self.transaction.hash())
+            .context("failed to verify transaction signature")
+    }
+
+    /// Returns the address which signed the transaction.
+    ///
+    /// # Errors
+    ///
+    /// - If the public key cannot be converted into an address
+    pub(crate) fn signer_address(&self) -> Result<Address> {
+        Address::try_from(&self.public_key)
+    }
+
+    /// Attempts to decode a signed transaction from the given bytes.
+    ///
+    /// # Errors
+    ///
+    /// - If the bytes cannot be decoded into the prost-generated `SignedTransaction` type
+    /// - If the transaction value is missing
+    /// - If the transaction value is not a valid transaction type (ie. does not correspond to any
+    ///   component)
+    /// - If the signature cannot be decoded
+    /// - If the public key cannot be decoded
+    pub(crate) fn try_from_slice(bytes: &[u8]) -> Result<Self> {
```

Comment:
I think the `ProtoSignedTransaction` should probably be re-exported since importing it from astria_proto every time might be annoying, what do you think? 

would renaming this to `try_from_proto_bytes` or something also be better?

Reply:
It might be annoying - but at the moment that is extremely rarely down so we don't get a lot from the extra brevity so I don't think this would cause a lot of churn (also because this is a private interface).

On the other hand separating deserialization from verification helps in tests and in understanding how the types moves in and out of our program.